From 5e29fb9362af35e4950022325bcac18001df58ea Mon Sep 17 00:00:00 2001
From: Alexander Sosedkin <asosedkin@redhat.com>
Date: Thu, 19 Mar 2026 19:33:02 +0100
Subject: [PATCH 1/3] x509/ocsp: check OCSP delegated signer EKU OID to full
 length

The retrieved certificate purpose OID was compared against the expected
1.3.6.1.5.5.7.3.9 value without checking if the lengths were identical,
allowing a value that constitutes a prefix to match.
The check now compares the length as well.

Reported-by: Joshua Rogers of AISLE Research Team <joshua@joshua.hu>
Co-authored-by: Joshua Rogers of AISLE Research Team <joshua@joshua.hu>
Signed-off-by: Alexander Sosedkin <asosedkin@redhat.com>
---
 lib/x509/ocsp.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lib/x509/ocsp.c b/lib/x509/ocsp.c
index caa511e9d..74fd19ace 100644
--- a/lib/x509/ocsp.c
+++ b/lib/x509/ocsp.c
@@ -2174,7 +2174,10 @@ static int check_ocsp_purpose(gnutls_x509_crt_t signercert)
 			return gnutls_assert_val(rc);
 		}
 
-		if (memcmp(oidtmp, GNUTLS_KP_OCSP_SIGNING, oidsize) != 0) {
+		/* x509_read_value() includes NUL in size for OIDs on 3.6,
+		unlike in 3.8 */
+		if (oidsize != sizeof(GNUTLS_KP_OCSP_SIGNING) ||
+		    memcmp(oidtmp, GNUTLS_KP_OCSP_SIGNING, oidsize) != 0) {
 			gnutls_assert();
 			continue;
 		}
-- 
2.53.0


